src/dotty/tools/dotc/ast/Positioned.scala

```diff
@@ -58,17 +58,43 @@ abstract class Positioned extends DotClass with Product {
    *  and transitively visit their children.
    */
   private def setChildPositions(pos: Position): Unit = {
```

Comment:
So what was wrong with the previous version of this method? And was it impossible to write the new version using recursive calls? I think it'd be helpful to add some comments explaining how this algorithm work and what are the invariants for all the mutable variables.